File: 1814-count-nice-pairs-in-an-array/1814-count-nice-pairs-in-an-array.cpp
class Solution {
public:
    
    typedef long long ll;
    const ll M = 1e9 + 7;
    
    ll rev(int num){
        ll number = 0LL;
        while(num){
            int rem = (num%10);
            number = (number*10LL) + (ll)rem;
            num/=10;
        }
        return number;
    }
    
    int countNicePairs(vector<int>& nums) {
        int n = nums.size();
        unordered_map<ll,int>freq;
        ll answer = 0LL;
        for(int i=0;i<n;i++){
            ll curr_contribution = (ll)nums[i] - rev(nums[i]);
            answer += freq[curr_contribution];
            answer %= M;
            freq[curr_contribution]++;
        }
        return answer;
    }
};

// nums[i] - rev(nums[i]) = nums[j] - rev(nums[j])
    
// 42-24 = 18
// 11-11 = 0
// 1-1   = 0    
// 97-79 = 18     

File: 1814-count-nice-pairs-in-an-array/1814-count-nice-pairs-in-an-array.java
class Solution {
    
    final long M = 1000000007;
    
    long rev(int num){
        long number = 0;
        while(num > 0){
            int rem = (num%10);
            number = (number*10L) + (long)rem;
            num /= 10;
        }
        return number;
    }
    
    public int countNicePairs(int[] nums) {
        int n = nums.length;
        long answer = 0L;
        HashMap<Long,Integer> freq = new HashMap<>();
        for(int i=0;i<n;i++){
            long curr_contribution = (long)nums[i] - rev(nums[i]);
            if(freq.containsKey(curr_contribution)){
                answer += freq.get(curr_contribution);
                answer %= M;
                freq.put(curr_contribution, freq.get(curr_contribution) + 1);
            }
            else{
                freq.put(curr_contribution, 1);
            }
        }
        return (int)answer;
    }
}